projects/mc_interface/src/motor_controller.c

```diff
@@ -0,0 +1,166 @@
+      can_data.drive_cmd.motor_current_percentage = storage->target_current_percentage;
+    } else {
+      // Velocity control (non-primary): use torque control instead, but copy the current setpoint
+      // from the primary motor controller
+      can_data.drive_cmd.motor_velocity_ms =
+          (storage->cruise_is_braking) ? 0.0f : WAVESCULPTOR_FORWARD_VELOCITY;
+      can_data.drive_cmd.motor_current_percentage = storage->cruise_current_percentage;
+    }
+    msg.data = can_data.raw;
+
+    generic_can_tx(storage->settings.can_uart, &msg);
+  }
+  storage->timeout_counter++;
+
+  soft_timer_start_millis(MOTOR_CONTROLLER_DRIVE_TX_PERIOD_MS, prv_periodic_tx, storage, NULL);
+}
+
+StatusCode motor_controller_init(MotorControllerStorage *controller,
+                                 const MotorControllerSettings *settings) {
+  memset(controller, 0, sizeof(*controller));
+  controller->settings = *settings;
+
+  WaveSculptorCanId can_id = {
+    .device_id = controller->settings.ids[0].motor_controller,
+    .msg_id = WAVESCULPTOR_MEASUREMENT_ID_BUS,
+  };
+
+  // Only bother registering the bus measurement handler for the first motor controller since
+  // that's all we care about
+  generic_can_register_rx(controller->settings.can_uart, prv_bus_measurement_rx,
+                          GENERIC_CAN_EMPTY_MASK, can_id.raw, false, controller);
+
+  for (size_t i = 0; i < NUM_MOTOR_CONTROLLERS; i++) {
+    can_id.device_id = controller->settings.ids[i].motor_controller;
+    can_id.msg_id = WAVESCULPTOR_MEASUREMENT_ID_VELOCITY;
+    generic_can_register_rx(controller->settings.can_uart, prv_velocity_measurement_rx,
+                            GENERIC_CAN_EMPTY_MASK, can_id.raw, false, controller);
+  }
+
+  return soft_timer_start_millis(MOTOR_CONTROLLER_DRIVE_TX_PERIOD_MS, prv_periodic_tx, controller,
+                                 NULL);
+}
+
+StatusCode motor_controller_set_speed_cb(MotorControllerStorage *controller,
+                                         MotorControllerSpeedCb speed_cb, void *context) {
+  controller->settings.speed_cb = speed_cb;
+  controller->settings.context = context;
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode motor_controller_set_throttle(MotorControllerStorage *controller, int16_t throttle,
+                                         EEDriveOutputDirection direction) {
+  // Use impossible target velocity for torque control
+  const float velocity_lookup[] = {
+    [EE_DRIVE_OUTPUT_DIRECTION_NEUTRAL] = 0.0f,
+    [EE_DRIVE_OUTPUT_DIRECTION_FORWARD] = WAVESCULPTOR_FORWARD_VELOCITY,
+    [EE_DRIVE_OUTPUT_DIRECTION_REVERSE] = WAVESCULPTOR_REVERSE_VELOCITY,
+  };
+
+  float target_velocity = velocity_lookup[direction];
+
+  if (direction == EE_DRIVE_OUTPUT_DIRECTION_NEUTRAL) {
+    // Neutral - coast
+    throttle = 0;
+  } else if (throttle < 0) {
+    // Braking - velocity = 0, brake %
+    throttle *= -1;
+    target_velocity = 0.0f;
+  }
+
+  // TODO(ELEC-338): may need critical section
```

Comment:
This should need one if this gets interrupted by a request to change state it might end up invalid...